toolchain/check/impl_lookup.cpp

```diff
@@ -204,29 +212,63 @@ auto LookupImplWitness(Context& context, SemIR::LocId loc_id,
   // TODO: Implement the termination rule, which requires looking at the
   // complexity of the types on the top of (or throughout?) the stack:
   // https://docs.carbon-lang.dev/docs/design/generics/details.html#termination-rule
-  for (auto entry : stack) {
+  for (auto [i, entry] : llvm::enumerate(stack)) {
     if (entry.type_const_id == type_const_id &&
         entry.interface_const_id == interface_const_id) {
+      auto facet_type_type_id =
+          context.types().GetTypeIdForTypeConstantId(interface_const_id);
       CARBON_DIAGNOSTIC(ImplLookupCycle, Error,
-                        "cycle found in lookup of interface {0} for type {1}",
-                        std::string, SemIR::TypeId);
-      context.emitter()
-          .Build(loc_id, ImplLookupCycle, "<TODO: interface name>",
-                 context.types().GetTypeIdForTypeConstantId(type_const_id))
-          .Emit();
+                        "cycle found in search for impl of {0} for type {1}",
+                        SemIR::TypeId, SemIR::TypeId);
+      auto builder = context.emitter().Build(
+          loc_id, ImplLookupCycle, facet_type_type_id,
+          context.types().GetTypeIdForTypeConstantId(type_const_id));
+      for (auto& active_entry : llvm::drop_begin(stack, i)) {
+        if (active_entry.impl_loc.has_value()) {
+          CARBON_DIAGNOSTIC(ImplLookupCycleNote, Note,
+                            "determining if this impl clause matches", );
+          builder.Note(active_entry.impl_loc, ImplLookupCycleNote);
+        }
+      }
+      builder.Emit();
       return SemIR::ErrorInst::SingletonInstId;
     }
   }
 
+  // The `interface_id` is the single interface in the `interface_const_id`
+  // facet type. In the future, a facet type may include more than a single
+  // interface, but for now that is unhandled with a TODO.
+  auto interface_id = [&] {
```

Comment:
Is this worth factoring out as a helper function? We do have a style preference for short functions; LookupImplWitness is now just shy of 100 LOC, and this looks like a good cut point. I'll admit, I also have a preference for avoiding lambdas when only used for flow control as here (since it's neither reused, nor a parameter): they make `return` harder to interpret when they get too long.

Reply:
Sure, I have also moved the cycle detection out to a helper function as well.